### PR TITLE
LRO with resource-names

### PR DESCRIPTION
**Pull request by chrisdunelm**

### CODE REVIEW

File: Google.Api.Generator.Tests/ProtoTests/Lro/Lro.proto [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  // Test an LRO RPC with a method signature that contains resource-names.
+  rpc ResourcedMethod(ResourceRequest) returns(google.longrunning.Operation) {
+    option (google.api.operation) = {
+      response_type: "LroResponse"
+      metadata_type: "LroMetadata"
+    };
+    option (google.api.method_signature) = {
+	  fields: ["name"]
+	};
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/Lro/Lro.proto]: Nit: outdent these two lines

File: Google.Api.Generator.Tests/ProtoTests/Lro/Lro.proto [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+service Lro {
+  option (google.api.default_host) = "lro.example.com";
+
+  // Test an LRO RPC with a method signature.
+  rpc SignatureMethod(Request) returns(google.longrunning.Operation) {
+    option (google.api.operation) = {
+      response_type: "LroResponse"
+      metadata_type: "LroMetadata"
+    };
+	option (google.api.method_signature) = {
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/Lro/Lro.proto]: Nit: indentation

File: Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+using scg = System.Collections.Generic;
+using sco = System.Collections.ObjectModel;
+using st = System.Threading;
+using stt = System.Threading.Tasks;
+
+namespace Testing.Lro
+{
+    public abstract class LroClient
+    {
+        public lro::Operation<LroResponse, LroMetadata> SignatureMethod(Request request, gaxgrpc::CallSettings callSettings) => null;
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs]: It looks like this is incomplete:

- Missing doc comments
- Missing virtual modifier
- Missing throwing implementation

File: Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+using grpcinter = Grpc.Core.Interceptors;
+using sys = System;
+using scg = System.Collections.Generic;
+using sco = System.Collections.ObjectModel;
+using st = System.Threading;
+using stt = System.Threading.Tasks;
+
+namespace Testing.Lro
+{
+    public abstract class LroClient
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs]: No LroClientImpl?

File: Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        /// Test an LRO RPC with a method signature that contains resource-names.
+        /// </summary>
+        /// <param name="name">
+        /// </param>
+        /// <param name="cancellationToken">A <see cref="st::CancellationToken"/> to use for this RPC.</param>
+        /// <returns>A Task containing the RPC response.</returns>
+        public virtual stt::Task<lro::Operation<LroResponse, LroMetadata>> ResourcedMethodAsync(ResourceName name, st::CancellationToken cancellationToken) =>
+            ResourcedMethodAsync(name, gaxgrpc::CallSettings.FromCancellationToken(cancellationToken));
+        // TEST_END
+    }
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs]: I seem to remember we used to generate a method to fetch an operation based on its name as well... is that coming in another PR?

File: Google.Api.Generator.Tests/ProtoTests/Lro/Lro.proto (hunk 2 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/Lro/Lro.proto]: Oops, done

File: Google.Api.Generator.Tests/ProtoTests/Lro/Lro.proto (hunk 1 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/Lro/Lro.proto]: done

File: Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs (hunk 4 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs]: These tests, and some of the other tests are badly named.
They are not testing the core LRO functionality, that's tested in `BasicLro`. This test is for advanced LRO functionality - in this case it's only testing that the method-signature and resource-name parts work properly.
So all the "normal" LRO code is deliberately missing from this test.

I'll refactor in a future PR to re-name the tests to be more descriptive.

File: Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs (hunk 3 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs]: This line of code is not part of the test, so is deliberately missing anything that isn't strictly required for the code to compile.

File: Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs (hunk 5 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/Lro/LroClient.cs]: See previous comment. The methods you mention are already tested in the `BasicLro` tests.

**Pull request merged**